### PR TITLE
Timeouts api refactor

### PR DESCRIPTION
**Pull request by Dreamlone**
Setting parameters via API has become a bit more clear

Also added a test for path compatibility during serialization. Slightly extended the list of initial assumptions for tabular data with linear models

Fixed a bug with metric evaluation

Change default 'max_depth' and 'max_arity'

### CODE REVIEW

File: fedot/api/api_utils/api_composer.py [hunk 1]
```diff
@@ -327,22 +289,66 @@ def tuner_metric_by_name(self, metric_name, train_data: InputData, task: Task):
...
+                            composer_requirements, pipeline_gp_composed,
+                            timer: ApiTime):
+        """ Launch tuning procedure for obtained pipeline by composer """
+        if tuning_params['with_tuning'] is False:
+            # Return source pipeline
+            return pipeline_gp_composed
+
+        timeout_for_tuning, iterations = timer.determine_resources_for_tuning()
+
+        if timeout_for_tuning < 15:
```
Comment (nicl-nno) [inline fedot/api/api_utils/api_composer.py:302]: Стоит наверное вынести в константы (мб вообще отдельный скрипт для них создать)

File: fedot/pattern_wrappers.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+def singleton(class_):
```
Comment (nicl-nno) [inline fedot/pattern_wrappers.py:1]: Может это не в корень, а в https://github.com/nccr-itmo/FEDOT/tree/master/fedot/utilities?

File: fedot/api/api_utils/api_composer.py (hunk 1 above)
Comment (Dreamlone) [reply fedot/api/api_utils/api_composer.py]: Так и сделал [тут](https://github.com/nccr-itmo/FEDOT/pull/587/commits/a234dda0e80b6a10b7dede67b0ebc54860006e3e#diff-4ef9b7a2a2edfb22283c66a7a8b1fc6a6e9c9f10387c9d4854aef6a729a00c69R1-R2), потом расширил следующим [коммитом](https://github.com/nccr-itmo/FEDOT/pull/587/commits/dca4f5bf7ecb300e363d76aa0238fe3d0a498ea5#diff-4ef9b7a2a2edfb22283c66a7a8b1fc6a6e9c9f10387c9d4854aef6a729a00c69R3-R5)

File: fedot/pattern_wrappers.py (hunk 2 above)
Comment (Dreamlone) [reply fedot/pattern_wrappers.py]: [Перенёс](https://github.com/nccr-itmo/FEDOT/pull/587/commits/a234dda0e80b6a10b7dede67b0ebc54860006e3e#)

File: fedot/api/api_utils/api_composer.py [hunk 3]
```diff
@@ -237,15 +233,19 @@ def compose_fedot_model(self, api_params: dict, composer_params: dict, tuning_pa
...
+        if timeout_were_set and init_pipeline_fit_time >= timer.datetime_composing / composer_params['pop_size']:
             api_params['logger'].message(f'Timeout is too small for composing '
-                                         f'because fit_time is {pipeline_fit_time.total_seconds()} sec.,'
+                                         f'because fit_time is {init_pipeline_fit_time.total_seconds()} sec.,'
                                          ' so it is skipped.')
+            # Use initial pipeline as final solution
             pipeline_gp_composed = fitted_initial_pipeline
         else:
+            # Launch pipeline structure composition
+            timer.start_composing()
```
Comment (IIaKyJIuH) [inline fedot/api/api_utils/api_composer.py:246]: Здесь тоже пригодился бы контекстный менеджер на таймере, тогда бы не пришлось в 
[`tune_final_pipeline(...)->determine_resources_for_tuning()`](https://github.com/nccr-itmo/FEDOT/pull/587/files#diff-bb87dc34890bf92350ee4d87a2488b24ce8b9e9e41f43cc36c35d65254135c5fR317) вызывать у него [`end_composing()`](https://github.com/nccr-itmo/FEDOT/pull/587/files#diff-cc7d8a57ff6077f9ea0a2f35f049483111dbf93c0753043d1c14f50b7b03af95R46)

File: fedot/api/time.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    def start_composing(self):
+        self.starting_time_for_composing = datetime.datetime.now()
+
+    def end_composing(self):
+        if self.starting_time_for_composing is None:
+            # Composing has not been started previously
+            return self.composing_spend_time
+
+        self.composing_spend_time = datetime.datetime.now() - self.starting_time_for_composing
+        return self.composing_spend_time
```
Comment (IIaKyJIuH) [inline fedot/api/time.py:52]: Точно ли здесь и в [`end_tuning()`](https://github.com/nccr-itmo/FEDOT/pull/587/files#diff-cc7d8a57ff6077f9ea0a2f35f049483111dbf93c0753043d1c14f50b7b03af95R57) понадобится возвращать значения соответствующих полей?

File: fedot/api/api_utils/params.py [hunk 5]
```diff
@@ -145,3 +153,35 @@ def get_task_params(problem, task_params: Optional[TaskParams] = None):
...
+        api_params['num_of_generations'] = num_of_generations
+    elif timeout > 0:
+        api_params['timeout'] = timeout
+        if num_of_generations is not None:
+            api_params['num_of_generations'] = num_of_generations
+        else:
+            api_params['num_of_generations'] = 10000
+    else:
+        raise ValueError(f'invalid "timeout" value: timeout={timeout}')
+    return api_params
```
Comment (IIaKyJIuH) [inline fedot/api/api_utils/params.py:174]: Я сам частично писал этот код и только сейчас заметил косяк.

Строки [165, 167, 169, 170] можно убрать, потому что там перезапись того же самого значения.

**Review by IIaKyJIuH — CHANGES_REQUESTED** [review]

File: test/data/pipeline_with_old_paths/pipeline_with_old_paths.json [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+			"nodes_from": [
+				1
+			],
+			"fitted_operation_path": "fitted_operations/operation_0.pkl",
+			"rating": null
+		}
+	],
+	"preprocessing": "preprocessing/data_preprocessor.pkl",
+	"descriptive_id": "(/n_scaling_default_params;)/n_ridge_default_params"
+}
```
[the remark this answers is not in the document]
Comment (nicl-nno) [reply test/data/pipeline_with_old_paths/pipeline_with_old_paths.json:33]: Для JSON-ов это вроде не специфицировано.

File: fedot/api/time.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+            return self.tuning_spend_time
+        self.tuning_spend_time = datetime.datetime.now() - self.starting_time_for_tuning
+        return self.tuning_spend_time
+
+    def determine_resources_for_tuning(self):
+        """
+        Based on time spend for composing and initial pipeline fit determine
+        how much time and how many iterations are needed for tuning
+        """
+        spended_time_for_composing = self.end_composing() + self.init_pipeline_fit_time
```
[the remark this answers is not in the document]
Comment (Dreamlone) [reply fedot/api/time.py:72]: Пока прикрыл вот [такой проверкой](https://github.com/nccr-itmo/FEDOT/pull/587/commits/805fb2f45a122787102350eadadd5221c025ae92#diff-cc7d8a57ff6077f9ea0a2f35f049483111dbf93c0753043d1c14f50b7b03af95R69-R71)
Хотя конечно не люблю самостоятельно выбрасывать ошибки, но здесь это поможет не напортачить при внесении изменений в API

File: fedot/api/time.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        # Time for tuning
+        self.starting_time_for_tuning = None
+        self.tuning_spend_time = 0
+
+        # Time to fit stand-alone pipeline
+        self.init_pipeline_fit_time = None
+
+    def __define_timeouts_for_stages(self):
+        """ Determine timeouts for tuning and composing """
+        if self.time_for_automl is None:
```
[the remark this answers is not in the document]
Comment (Dreamlone) [reply fedot/api/time.py]: Согласен, [сделал](https://github.com/nccr-itmo/FEDOT/pull/587/commits/805fb2f45a122787102350eadadd5221c025ae92#diff-cc7d8a57ff6077f9ea0a2f35f049483111dbf93c0753043d1c14f50b7b03af95L34-L35)

File: fedot/api/api_utils/api_composer.py [hunk 9]
```diff
@@ -237,15 +233,19 @@ def compose_fedot_model(self, api_params: dict, composer_params: dict, tuning_pa
                                          initial_assumption=api_params['initial_assumption'],
                                          logger=api_params['logger'])
         gp_composer = None
-        if pipeline_fit_time >= timeout_for_composing / composer_params['pop_size']:
+        timeout_were_set = timer.datetime_composing not in [-1, None]
```
[the remark this answers is not in the document]
Comment (Dreamlone) [reply fedot/api/api_utils/api_composer.py]: Учел, [исправил](https://github.com/nccr-itmo/FEDOT/pull/587/commits/805fb2f45a122787102350eadadd5221c025ae92#diff-bb87dc34890bf92350ee4d87a2488b24ce8b9e9e41f43cc36c35d65254135c5fL237-L238)

File: fedot/api/api_utils/constants.py [hunk 10]
```diff
@@ -0,0 +1,5 @@
+MINIMUM_SECONDS_FOR_TUNING = 15
+DEFAULT_NUMBER_TUNING_ITERATIONS = 1000
+DEFAULT_TIMEOUT_API_MINUTES = 5.0
```
[the remark this answers is not in the document]
Comment (Dreamlone) [reply fedot/api/api_utils/constants.py]: Ну ты завернул
[Исправил](https://github.com/nccr-itmo/FEDOT/pull/587/commits/805fb2f45a122787102350eadadd5221c025ae92#diff-4ef9b7a2a2edfb22283c66a7a8b1fc6a6e9c9f10387c9d4854aef6a729a00c69L1-R3)

File: test/data/pipeline_with_old_paths/pipeline_with_old_paths.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (Dreamlone) [reply test/data/pipeline_with_old_paths/pipeline_with_old_paths.json:33]: Не стоит. Как FEDOT сериализует модели, так лучше и оставлять кмк

File: fedot/api/api_utils/params.py (hunk 5 above)
Comment (Dreamlone) [reply fedot/api/api_utils/params.py]: [Переделал](https://github.com/nccr-itmo/FEDOT/pull/587/commits/805fb2f45a122787102350eadadd5221c025ae92#diff-35c9039b23102036436d1536f49036a1cdfa7214c697fbd184e84071246dead8L165-L173)

**Review by v1docq — COMMENTED** [review]
Small review

File: fedot/api/api_utils/api_composer.py [hunk 11]
```diff
@@ -7,6 +7,7 @@
 from deap import tools
 from sklearn.metrics import mean_squared_error, roc_auc_score as roc_auc
 
+from fedot.api.api_utils.constants import MINIMAL_SECONDS_FOR_TUNING, DEFAULT_TUNING_ITERATIONS_NUMBER
```
[the remark this answers is not in the document]
Comment (Dreamlone) [reply fedot/api/api_utils/api_composer.py]: По заглавному шрифту, ага, действуем по стандартному [соглашению по неймингу](https://pythonz.net/references/named/constants/)

По поводу всех констант вполне может быть резонно. Можно подготовить этот файл где-то рядом с `core`. Собственно, так и сделал в этом [коммите](https://github.com/nccr-itmo/FEDOT/pull/587/commits/1739a40967df5872ab0a5ade952826ff4ddefafd). Потом можем с течением времени туда сносить подобные константы

File: fedot/api/api_utils/params.py [hunk 12]
```diff
@@ -104,13 +90,35 @@ def change_preset_for_label_encoded_data(self, task: Task):
...
+            self.task_params = TsForecastingParams(forecast_length=DEFAULT_FORECAST_LENGTH)
+
+        if input_params['problem'] == 'clustering':
+            raise ValueError('This type of task is not not supported in API now')
+
     @staticmethod
     def get_default_evo_params(problem: str):
         """ Dictionary with default parameters for composer """
-        params = {'max_depth': 3,
-                  'max_arity': 4,
```
[the remark this answers is not in the document]
Comment (Dreamlone) [reply fedot/api/api_utils/params.py:111]: Согласен, [изменил](https://github.com/nccr-itmo/FEDOT/pull/587/commits/1739a40967df5872ab0a5ade952826ff4ddefafd#diff-35c9039b23102036436d1536f49036a1cdfa7214c697fbd184e84071246dead8L118-R119)

File: fedot/api/time.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Dreamlone) [reply fedot/api/time.py]: Update, решил просто вынести init_pipeline_fit_time как аргумент в метод determine_resources_for_tuning. Пришлось его протащить через метод в API, зато теперь исключена (я надеюсь), проблема с тем, что init_pipeline_fit_time будет None

File: fedot/api/api_utils/api_composer.py [hunk 13]
```diff
@@ -343,22 +306,66 @@ def tuner_metric_by_name(self, metric_name, train_data: InputData, task: Task):
...
+            timeout_for_tuning = abs(timeout_for_tuning) / 60
+            pipeline_gp_composed = pipeline_gp_composed.fine_tune_all_nodes(loss_function=tuner_loss,
+                                                                            loss_params=loss_params,
+                                                                            input_data=api_params['train_data'],
+                                                                            iterations=DEFAULT_NUMBER_TUNING_ITERATIONS,
+                                                                            timeout=timeout_for_tuning,
+                                                                            cv_folds=folds,
+                                                                            validation_blocks=vb_number)
+            api_params['logger'].message('Hyperparameters tuning finished')
+            timer.end_tuning()
```
[the remark this answers is not in the document]
Comment (Dreamlone) [reply fedot/api/api_utils/api_composer.py]: Решил запускать в with отдельно взятые методы, например, как [здесь](https://github.com/nccr-itmo/FEDOT/pull/587/commits/1739a40967df5872ab0a5ade952826ff4ddefafd#diff-bb87dc34890bf92350ee4d87a2488b24ce8b9e9e41f43cc36c35d65254135c5fR339-R352)

File: fedot/api/api_utils/api_composer.py (hunk 3 above)
Comment (Dreamlone) [reply fedot/api/api_utils/api_composer.py]: Согласен, [исправил](https://github.com/nccr-itmo/FEDOT/pull/587/commits/1739a40967df5872ab0a5ade952826ff4ddefafd#diff-bb87dc34890bf92350ee4d87a2488b24ce8b9e9e41f43cc36c35d65254135c5fL246-R251)

File: fedot/api/api_utils/api_composer.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (v1docq) [reply fedot/api/api_utils/api_composer.py]: Тогда держим константы в одном месте

**Review by v1docq — APPROVED** [review]
С изменениями согласен

File: fedot/api/time.py (hunk 4 above)
Comment (Dreamlone) [reply fedot/api/time.py]: Нет, [убрал этот момент](https://github.com/nccr-itmo/FEDOT/pull/587/commits/1739a40967df5872ab0a5ade952826ff4ddefafd#diff-cc7d8a57ff6077f9ea0a2f35f049483111dbf93c0753043d1c14f50b7b03af95R41-R48)

**Review by valer1435 — APPROVED** [review]

File: fedot/api/api_utils/constants.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (IIaKyJIuH) [reply fedot/api/api_utils/constants.py]: Соре, само вырвалось)

File: test/data/pipeline_with_old_paths/pipeline_with_old_paths.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (IIaKyJIuH) [reply test/data/pipeline_with_old_paths/pipeline_with_old_paths.json:33]: Просто заметил, что GitHub ругается на отсутствие пустой строки.

Не принципиально, но может стоит тогда вообще сделать issue по сериализации моделей с добавлением пустой строки?

[the remark this answers is not in the document]
Comment (Dreamlone) [reply test/data/pipeline_with_old_paths/pipeline_with_old_paths.json:33]: Мне кажется, не стоит. Коля верно написал, что для не .py по стандарту PEP8 не предусмотрено подобных стилистических правил. А на gihub лучше в этом плане не смотреть, главное, чтобы PEP8 бот не ругался)

**Review by ChrisLisbon — APPROVED** [review]

File: fedot/api/time.py [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+    @property
+    def datetime_composing(self):
+        if self.timeout_for_composing is None:
+            return None
+        return datetime.timedelta(minutes=self.timeout_for_composing)
+
+    def __exit__(self, *args):
+        self.log.info(f'Composition time: {round(self.minutes_from_start, 3)} min')
+        if self.process_terminated:
+            self.log.info('Algorithm was terminated due to processing time limit')
```
Comment (IIaKyJIuH) [inline fedot/api/time.py:88]: Лишний метод

File: test/data/pipeline_with_old_paths/pipeline_with_old_paths.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (IIaKyJIuH) [reply test/data/pipeline_with_old_paths/pipeline_with_old_paths.json:33]: Тогда всё отлично)

File: fedot/api/api_utils/api_composer.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (IIaKyJIuH) [reply fedot/api/api_utils/api_composer.py]: Красиво получилось, молодец

File: fedot/api/time.py [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+    @contextmanager
+    def launch_composing(self):
+        """ Wrap composing process with timer """
+        self.starting_time_for_composing = datetime.datetime.now()
+        yield
+        if self.starting_time_for_composing is None:
+            # Composing has not been started previously
+            return self.composing_spend_time
+
+        self.composing_spend_time = datetime.datetime.now() - self.starting_time_for_composing
```
Comment (IIaKyJIuH) [inline fedot/api/time.py:50]: Полагаю, теперь часть с "return" не нужна, а на её место можно поставить 50 строку, взяв обратное условие.

Точно так же можно поступить и с [`launch_tuning()`](https://github.com/nccr-itmo/FEDOT/pull/587/files#diff-cc7d8a57ff6077f9ea0a2f35f049483111dbf93c0753043d1c14f50b7b03af95R53)

**Review by IIaKyJIuH — CHANGES_REQUESTED** [review]

File: fedot/api/time.py (hunk 14 above)
Comment (Dreamlone) [reply fedot/api/time.py]: Да, чего то я тут погорячился, [убрал](https://github.com/nccr-itmo/FEDOT/pull/587/commits/6d085ee2d52f7f2ed1f45d564a37fc74fefea6c9)

File: fedot/api/time.py (hunk 15 above)
Comment (Dreamlone) [reply fedot/api/time.py]: Согласен, даже проще, можно просто убрать блок с проверкой "if", так как self.starting_time_for_composing None точно быть не может, так как запускается прямо в этом же методе
[Коммит](https://github.com/nccr-itmo/FEDOT/pull/587/commits/6d085ee2d52f7f2ed1f45d564a37fc74fefea6c9)

Comment (IIaKyJIuH) [reply fedot/api/time.py]: Ты прав

**Review by IIaKyJIuH — APPROVED** [review]
